core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -40,38 +41,70 @@ internal actual fun <T> createParametrizedCache(factory: (KClass<Any>, List<KTyp
 }
 
-private class ClassValueCache<T>(compute: (KClass<*>) -> KSerializer<T>?) : SerializerCache<T> {
-    private val classValue = ClassValueWrapper(compute)
+private class ClassValueCache<T>(val compute: (KClass<*>) -> KSerializer<T>?) : SerializerCache<T> {
+    private val classValue = ClassValueReferences<CacheEntry<T>>()
 
-    override fun get(key: KClass<Any>): KSerializer<T>? = classValue[key.java].serializer
+    override fun get(key: KClass<Any>): KSerializer<T>? {
+        return classValue
+            .getOrSet(key.java) { CacheEntry(compute(key)) }
+            .serializer
+    }
 }
 
+/**
+ * A class that combines the capabilities of ClassValue and SoftReference.
+ * Softly binds the calculated value to the specified class.
+ *
+ * [SoftReference] used to prevent class loaders from leaking,
+ * since the value can transitively refer to an instance of type [Class], this may prevent the loader from
+ * being collected during garbage collection.
+ *
+ * In the first calculation the value is cached, every time [getOrSet] is called, a pre-calculated value is returned.
+ *
+ * However, the value can be collected during garbage collection (thanks to [SoftReference])
+ * - in this case, when trying to call the [getOrSet] function, the value will be calculated again and placed in the cache.
+ *
+ * An important requirement for a function generating a value is that it must be stable, so that each time it is called for the same class, the function returns similar values.
+ * In the case of serializers, these should be instances of the same class filled with equivalent values.
+ */
 @SuppressAnimalSniffer
-private class ClassValueWrapper<T>(private val compute: (KClass<*>) -> KSerializer<T>?): ClassValue<CacheEntry<T>>() {
-    /*
-     * Since during the computing of the value for the `ClassValue` entry, we do not know whether a nullable
-     *  serializer is needed, so we may need to differentiate nullable/non-null  caches by a level higher
-     */
-    override fun computeValue(type: Class<*>): CacheEntry<T> {
-        return CacheEntry(compute(type.kotlin))
+private class ClassValueReferences<T> : ClassValue<MutableSoftReference<T>>() {
+    override fun computeValue(type: Class<*>): MutableSoftReference<T> {
+        return MutableSoftReference()
+    }
+
+    inline fun getOrSet(key: Class<*>, crossinline factory: () -> T): T {
+        val ref: MutableSoftReference<T> = get(key)
+
+        ref.reference.get()?.let { return it }
+        // go to the slow path and create serializer with blocking, also wrap factory block
+        return createAndSet(ref) { factory() }
     }
-}
 
-private class ClassValueParametrizedCache<T>(private val compute: (KClass<Any>, List<KType>) -> KSerializer<T>?) : ParametrizedSerializerCache<T> {
-    private val classValue = ParametrizedClassValueWrapper<T>()
+    @Synchronized
+    private fun createAndSet(ref: MutableSoftReference<T>, factory: () -> T): T {
```

Comment:
Shouldn't we re-check the value of `reference` in the beginning of the method again?
It looks like the following situation is possible:
Thread1 — sees empty reference, goes into createAndSet
Thread2 — also sees empty reference (because it wasn't yet written by thread1), waits until the lock on createAndSet is available, and also creates a new value, overriding the value from thread1.

It's not that bad because all references are the same, and the discarded reference from thread1 will likely be collected soon, but the check is also easy to add

Reply:
Yes, my bad, I deleted an additional check due to code rework  :(